libs/mctl/mctl.c

```diff
@@ -0,0 +1,255 @@
+/*
+ * Phoenix-Pilot
+ *
+ * mctl.h - motors control module
+ * 
+ * Control over engines, arming and disarmig procedures. No thread safety imposed.
+ *
+ * Copyright 2022 Phoenix Systems
+ * Author: Mateusz Niewiadomski
+ *
+ * This file is part of Phoenix-Pilot software
+ *
+ * %LICENSE%
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <fcntl.h>
+#include <math.h>
+
+#include <board_config.h>
+
+#include "mctl.h"
+
+
+#define THROTTLE_DOWN   0.0    /* default/init/lowest position of throttle */
+#define THROTTLE_SCALER 100000 /* base thrtl->pwm scaling factor */
+
+struct {
+	FILE **motorFile;           /* motors pwm files descriptors */
+	float *mThrtl;              /* motors current throttle value */
+	unsigned int mctlInitFlag;  /* motors descriptors initialization flag */
+	unsigned int mctlArmedFlag; /* motors armed/disarmed flag */
+	unsigned int mots;          /* number of motors */
+} mctl_common;
+
+
+static int mctl_motWrite(unsigned int motorIdx, float thrtl)
+{
+	unsigned int thrtlVal;
+
+	if (motorIdx >= mctl_common.mots) {
+		return -1;
+	}
+
+	if (thrtl > 1.f) {
+		thrtl = 1.f;
+	}
+	else if (thrtl < 0.f) {
+		thrtl = 0.f;
+	}
+
+	thrtlVal = (unsigned int)((thrtl + 1.0f) * (float)THROTTLE_SCALER);
+
+	if (fprintf(mctl_common.motorFile[motorIdx], "%u\n", thrtlVal) < 0) {
+		fprintf(stderr, "mctl: cannot set PWM for motor: %d\n", motorIdx);
+		return -1;
+	}
+	fflush(mctl_common.motorFile[motorIdx]);
+
+	mctl_common.mThrtl[motorIdx] = thrtl;
+
+	return 0;
+}
+
+
+static inline int mctl_motOff(unsigned int motorIdx)
+{
+	/* if index out of bounds or error during write, return -1 */
+	return (motorIdx >= mctl_common.mots || fprintf(mctl_common.motorFile[motorIdx], "0") < sizeof("0") - 1) ? -1 : 0;
+}
+
+
+static inline void mctl_printRed(const char *msg)
+{
+	printf("%s%s%s", "\033[1;31m", msg, "\033[0m");
+	fflush(stdout);
+}
+
+
+int mctl_thrtlSet(unsigned int motorIdx, float targetThrottle, enum thrtlTempo tempo)
+{
+	float currThrtl, change, uchange, rate;
+	unsigned int steps;
+
+	if (!mctl_common.mctlInitFlag || !mctl_common.mctlArmedFlag) {
```

Comment:
To be compliant with MISRA rules:
```suggestion
	if (mctl_common.mctlInitFlag == 0 || mctl_common.mctlArmedFlag == 0) {
```

Reply:
As these are strict boolean flags `bool` type was used